### PR TITLE
MULE-10271: Revamp validation module based on mule 4 error handling c…

### PR DESCRIPTION
**Pull request by estebanwasinger**
…apabilities

### CODE REVIEW

File: extensions/validation/src/main/java/org/mule/extension/validation/api/ValidationErrorTypes.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.mule.runtime.extension.api.error.ErrorTypeDefinition;
+
+/**
+ * List of {@link ErrorTypeDefinition} that throws the {@link ValidationExtension}
+ *
+ * @since 4.0
+ */
+public enum ValidationErrorTypes implements ErrorTypeDefinition<ValidationErrorTypes> {
+  /**
```
Comment (juandesi) [inline extensions/validation/src/main/java/org/mule/extension/validation/api/ValidationErrorTypes.java]: empty line

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/exception/ModuleExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+   * The {@link ErrorType} will be taken from the {@link ErrorTypeRepository} using the
+   * {@link ErrorTypeDefinition} from the {@link ModuleException}
+   * </li>
+   * </ul>
+   *
+   * @param throwable to process
+   */
+  public Throwable processException(Throwable throwable) {
+    HashMap<byte[], byte[]> objectObjectHashMap = new HashMap<>();
+    objectObjectHashMap.put("Hola".getBytes(), "value".getBytes());
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/exception/ModuleExceptionHandler.java]: Hi

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/exception/ModuleExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  private Throwable handleTypedException(Throwable exception, ErrorTypeDefinition errorDefinition) {
+    if (isAllowedError(errorDefinition)) {
+      Optional<ErrorType> errorType = typeRepository.lookupErrorType(new ComponentIdentifier.Builder()
+          .withNamespace(extensionNamespace)
+          .withName(errorDefinition.getType())
+          .build());
+
+      if (errorType.isPresent()) {
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/exception/ModuleExceptionHandler.java]: orElseThrow on lookup

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/exception/ModuleExceptionHandler.java (hunk 2 above)
Comment (estebanwasinger) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/exception/ModuleExceptionHandler.java]: Alo

File: extensions/validation/src/main/java/org/mule/extension/validation/api/ObjectSource.java [hunk 4]
```diff
@@ -22,21 +23,23 @@
...
- * <p/>
+ * <p>
  * Instances of this class are to be considered thread-safe and reusable.
  *
  * @param <T> the type of the object to be returned
  * @since 3.7.0
  */
+//TODO MULE-11144 : Add '(isOneRequired = true)' when the bug is fixed
+@ExclusiveOptionals
 public class ObjectSource<T> {
```
Comment (marianogonzalez) [inline extensions/validation/src/main/java/org/mule/extension/validation/api/ObjectSource.java]: this class makes no sense anymore. If only the messageFactory is using it, then just move these parameters to that class

File: extensions/validation/src/main/java/org/mule/extension/validation/internal/CommonValidationOperations.java [hunk 5]
```diff
@@ -78,6 +83,7 @@ public void isFalse(boolean expression, @ParameterGroup(ERROR_GROUP) ValidationO
    * @param event the current {@link Event
    * @param config the current {@link ValidationExtension} that serves as config
    */
+  @Throws(ValidationErrorTypeProvider.class)
```
Comment (marianogonzalez) [inline extensions/validation/src/main/java/org/mule/extension/validation/internal/CommonValidationOperations.java]: can't this be declared at a more global level? looks like it should

Comment (estebanwasinger) [inline extensions/validation/src/main/java/org/mule/extension/validation/internal/CommonValidationOperations.java]: Requires support from SDK side

File: extensions/validation/src/main/java/org/mule/extension/validation/api/ObjectSource.java (hunk 4 above)
Comment (estebanwasinger) [inline extensions/validation/src/main/java/org/mule/extension/validation/api/ObjectSource.java]: This class is not used by the message factory, is used to create instances of custom validators

**Pull request merged**